// @fugle/trade/lib/order.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Order = void 0;
const enums_1 = require("./enums");
const PAYLOAD = Symbol('Order#payload');
class Order {
    constructor(payload) {
        this[PAYLOAD] = payload || {};
    }
    get payload() {
        return this[PAYLOAD];
    }
    setSymbol(symbol) {
        this.payload.stockNo = symbol;
        return this;
    }
    setSide(side) {
        this.payload.buySell = side;
        return this;
    }
    setPrice(price) {
        this.payload.price = price;
        return this;
    }
    setQuantity(quantity) {
        this.payload.quantity = quantity;
        return this;
    }
    setApCode(apCode) {
        this.payload.apCode = apCode;
        return this;
    }
    setPriceFlag(priceFlag) {
        this.payload.priceFlag = priceFlag;
        return this;
    }
    setBsFlag(bsFlag) {
        this.payload.bsFlag = bsFlag;
        return this;
    }
    setTradeType(tradeType) {
        this.payload.trade = tradeType;
        return this;
    }
    toObject() {
        return {
            stockNo: String(this.payload.stockNo),
            buySell: String(this.payload.buySell),
            price: String(this.payload.price || ''),
            quantity: String(this.payload.quantity),
            apCode: String(this.payload.apCode),
            priceFlag: String(this.payload.priceFlag),
            bsFlag: String(this.payload.bsFlag),
            trade: String(this.payload.trade),
        };
    }
}
exports.Order = Order;
Order.Side = enums_1.Side;
Order.ApCode = enums_1.ApCode;
Order.PriceFlag = enums_1.PriceFlag;
Order.BsFlag = enums_1.BsFlag;
Order.Trade = enums_1.TradeType;
